### PR TITLE
add utils to upload file, upload/download folder

### PR DESCRIPTION
**Pull request by fvankrieken**
A couple points

- do we want to use the `rich` printing/progress I copied over from data library? Essentially all `upload_file` is is a wrapper around boto3 client built-in upload with this progress. If we like this we could do a similar thing for download
- download_folder is essentially ported over from @DeaBardhoshi and @athursland 's work in db-checkbook
- ~~should acl have default? This is what we had in data library but it seems maybe it's good to make ourselves be explicit if we're making things being pushed public rather than defaulting to it~~ going with no default - acl must be passed explicitly. We could have a small wrapper in `recipes` if we really want one that defaults to public but this will likely be used to push publishing data as well
- ~~I have a funny mish-mash of os.path stuff and pathlib.Path~~ better now I think

### CODE REVIEW

File: dcpy/connectors/s3.py [hunk 1]
```diff
@@ -8,9 +18,83 @@ def client(
...
+
+
+## adapted from data library
+## https://github.com/NYCPlanning/db-data-library/blob/main/library/s3.py#L37
+def upload_file(
+    bucket: str,
+    path: str,
+    key: str,
+    acl: str = "public-read",
+    metadata: dict = {},
```
Comment (alexrichey) [inline dcpy/connectors/s3.py:38]: So this is a bizarre Python'ism (thanks Guido) but you'd typically pass None here due to [dangerous-default-values](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/dangerous-default-value.html

Comment (alexrichey) [reply dcpy/connectors/s3.py:38]: also, if you want to put the context above the fn in the docstring, that'd be 👍

Comment (fvankrieken) [reply dcpy/connectors/s3.py:38]: Good to know

**Comment by fvankrieken** [comment]
Made a couple tweaks around paths and reformatted - more ready for review now. I'll add any more changes in additional commits, will likely just squash this one

File: dcpy/connectors/s3.py [hunk 2]
```diff
@@ -8,9 +18,87 @@ def client(
...
+    return response
+
+
+def download_folder(
+    bucket, prefix, export_path, include_prefix_in_export=False
+) -> None:
+    """given bucket, prefix filter, and export path, download contents of folder from s3 recursively"""
+    export_path = Path(export_path)
+    if prefix[-1] != "/":
+        raise NotADirectoryError("prefix must be a folder path, ending with '/'")
```
Comment (alexrichey) [inline dcpy/connectors/s3.py:70]: Maybe I'm missing where we defined this Exception? Also, you might just use an assert here, unless you were planning to specifically handle `NotADirectoryError`s somewhere?

File: dcpy/connectors/s3.py [hunk 3]
```diff
@@ -8,9 +18,87 @@ def client(
...
+            ExtraArgs={"ACL": acl, "Metadata": metadata},
+            Callback=lambda complete: progress.update(task, completed=complete),
+        )
+    return response
+
+
+def download_folder(
+    bucket, prefix, export_path, include_prefix_in_export=False
+) -> None:
+    """given bucket, prefix filter, and export path, download contents of folder from s3 recursively"""
```
Comment (alexrichey) [inline dcpy/connectors/s3.py]: So does this just download a folder and it's non-folder children? Recursively makes me think you'd be recursing to download a full folder tree? (seems fine either way... maybe safer the first way, so you don't accidentally download a full bucket)

Comment (alexrichey) [reply dcpy/connectors/s3.py]: for later readers, yes, it is recursive because of how prefixes work

File: dcpy/connectors/s3.py (hunk 2 above)
Comment (fvankrieken) [reply dcpy/connectors/s3.py:70]: It's a built-in python exception, just figured give the most specific error possible. I'm personally fine with asserts as well, just depends a bit what we like for stylistic norms around errors

Comment (alexrichey) [reply dcpy/connectors/s3.py:70]: didn't realize it was a built-in. Seems fine!

File: dcpy/connectors/s3.py (hunk 3 above)
Comment (fvankrieken) [reply dcpy/connectors/s3.py]: Chatted in person but for posterity (aka @damonmcc ) - list_objects does find ALL objects with the given path prefix. So downloading a whole bucket (or first 1000, believe is default for list_objects) is certainly possible.

File: dcpy/connectors/s3.py [hunk 4]
```diff
@@ -8,9 +18,88 @@ def client(
...
+            if include_prefix_in_export:
+                export_path = export_path / prefix
+            if not (export_path / key_directory).exists():
+                os.makedirs(export_path / key_directory)
+            client_.download_file(bucket, obj["Key"], export_path / key)
+
+
+def upload_folder(
+    bucket: str,
+    local_folder_path: str,
```
Comment (alexrichey) [inline dcpy/connectors/s3.py:89]: do you think it makes sense just ask for `pathlib.Path`s in these methods?

Comment (fvankrieken) [reply dcpy/connectors/s3.py:89]: Yeah I went back and forth a little myself. I assume you think yes?

**Comment by fvankrieken** [comment]
download_folder is currently a tad broken, hold off on further reviews

**Comment by fvankrieken** [comment]
Functions working as expected! All set from my end

**Review by alexrichey — APPROVED** [review]
Good stuff!

**Comment by alexrichey** [comment]
Oh, one other thing. Can you add `rich` to requirements.txt/in?

**Pull request merged**